File: Animated Button/[C#][VB]-Animated Button/C# and VB/Animated Button/Animated Button.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Animate
{
    public partial class AnimatedButton : Button
    {
        #region Frame

        private class Frame
        {
            public Frame(int FrameCount, int CurrentFrame, Size FrameSize, Image ImageStrip)
            {
                this.FrameCount = FrameCount;
                this.CurrentFrame = CurrentFrame;
                this.FrameSize = FrameSize;
                this.ImageStrip = ImageStrip;
            }

            public Image ImageStrip
            {
                get;
                set;
            }
            public int FrameCount
            {
                get;
                set;
            }
            public int CurrentFrame
            {
                get;
                set;
            }
            public Size FrameSize
            {
                get;
                set;
            }
        }

        #endregion

        #region MouseOver

        private bool _IsMouseOver = false;
        private Frame MouseOverFrame
        {
            get;
            set;
        }
        private bool IsMouseOver
        {
            get
            {
                return _IsMouseOver;
            }
            set
            {
                _IsMouseOver = value;

                Refresh();
            }
        }
        protected override void OnMouseEnter(EventArgs e)
        {
            base.OnMouseEnter(e);

            IsMouseOver = true;
        }
        protected override void OnMouseLeave(EventArgs e)
        {
            base.OnMouseLeave(e);

            IsMouseOver = false;
        }

        #endregion

        #region MouseClick

        private bool _IsMouseDown = false;
        private bool IsMouseDown
        {
            get
            {
                return _IsMouseDown;
            }
            set
            {
                _IsMouseDown = value;

                Refresh();
            }
        }
        private Frame MouseClickFrame
        {
            get;
            set;
        }
        protected override void OnMouseDown(MouseEventArgs e)
        {
            base.OnMouseDown(e);

            IsMouseDown = true;
        }
        protected override void OnMouseUp(MouseEventArgs e)
        {
            base.OnMouseUp(e);

            IsMouseDown = false;
        }

        #endregion

        public AnimatedButton()
        {
            InitializeComponent();

            DoubleBuffered = true;

            MouseOverFrame = new Frame(25, 0, new Size(100, 30), Properties.Resources.ButtonMouseOver);
            MouseClickFrame = new Frame(25, 0, new Size(100, 30), Properties.Resources.ButtonMouseDown);

            this.Size = MouseOverFrame.FrameSize;
        }

        protected override void OnPaint(PaintEventArgs pe)
        {
            base.OnPaint(pe);

            pe.Graphics.Clear(BackColor); 

            pe.Graphics.InterpolationMode = System.Drawing.Drawing2D.InterpolationMode.HighQualityBicubic;
            pe.Graphics.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;
            pe.Graphics.CompositingQuality = System.Drawing.Drawing2D.CompositingQuality.HighQuality;

            if (!IsMouseOver && !IsMouseDown && MouseOverFrame.CurrentFrame == 0)
            {
                pe.Graphics.DrawImage(MouseOverFrame.ImageStrip, 
                    new Rectangle(0, 0, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height), 
                    new Rectangle(0, 0, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height), 
                    GraphicsUnit.Pixel);
            }
            else if (IsMouseOver && !IsMouseDown && MouseOverFrame.CurrentFrame != MouseOverFrame.FrameCount)
            {
                pe.Graphics.DrawImage(MouseOverFrame.ImageStrip, 
                    new Rectangle(0, 0, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height), 
                    new Rectangle(0, MouseOverFrame.CurrentFrame++ * MouseOverFrame.FrameSize.Height, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height), 
                    GraphicsUnit.Pixel);
            }
            else if (IsMouseOver && !IsMouseDown && MouseOverFrame.CurrentFrame == MouseOverFrame.FrameCount && MouseClickFrame.CurrentFrame == 0)
            {
                pe.Graphics.DrawImage(MouseOverFrame.ImageStrip, 
                    new Rectangle(0, 0, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height), 
                    new Rectangle(0, (MouseOverFrame.CurrentFrame - 1) * MouseOverFrame.FrameSize.Height, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height), 
                    GraphicsUnit.Pixel);
            }
            else if (IsMouseOver && IsMouseDown && MouseClickFrame.CurrentFrame != MouseClickFrame.FrameCount)
            {
                pe.Graphics.DrawImage(MouseClickFrame.ImageStrip, 
                    new Rectangle(0, 0, MouseClickFrame.FrameSize.Width, MouseClickFrame.FrameSize.Height), 
                    new Rectangle(0, MouseClickFrame.CurrentFrame++ * MouseClickFrame.FrameSize.Height, MouseClickFrame.FrameSize.Width, MouseClickFrame.FrameSize.Height), 
                    GraphicsUnit.Pixel);
            }
            else if (IsMouseOver && IsMouseDown && MouseClickFrame.CurrentFrame == MouseClickFrame.FrameCount)
            {
                pe.Graphics.DrawImage(MouseClickFrame.ImageStrip,
                    new Rectangle(0, 0, MouseClickFrame.FrameSize.Width, MouseClickFrame.FrameSize.Height), 
                    new Rectangle(0, (MouseClickFrame.CurrentFrame - 1) * MouseClickFrame.FrameSize.Height, MouseClickFrame.FrameSize.Width, MouseClickFrame.FrameSize.Height), 
                    GraphicsUnit.Pixel);
            }
            else if (IsMouseOver && !IsMouseDown && MouseClickFrame.CurrentFrame != 0)
            {
                pe.Graphics.DrawImage(MouseClickFrame.ImageStrip, 
                    new Rectangle(0, 0, MouseClickFrame.FrameSize.Width, MouseClickFrame.FrameSize.Height), 
                    new Rectangle(0, --MouseClickFrame.CurrentFrame * MouseClickFrame.FrameSize.Height, MouseClickFrame.FrameSize.Width, MouseClickFrame.FrameSize.Height), 
                    GraphicsUnit.Pixel);
            }
            else if (!IsMouseOver && !IsMouseDown && MouseOverFrame.CurrentFrame != 0)
            {
                pe.Graphics.DrawImage(MouseOverFrame.ImageStrip, 
                    new Rectangle(0, 0, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height), 
                    new Rectangle(0, --MouseOverFrame.CurrentFrame * MouseOverFrame.FrameSize.Height, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height), 
                    GraphicsUnit.Pixel);
            }

            using (StringFormat sf = new StringFormat())
            {
                sf.Alignment = StringAlignment.Center;
                sf.LineAlignment = StringAlignment.Center;

                pe.Graphics.DrawString(Text, Font, new SolidBrush(ForeColor), pe.ClipRectangle, sf);
            }
        }

        private void _FPS_Tick(object sender, EventArgs e)
        {
            Refresh();
        }
    }
}

File: Animated Button/[C#][VB]-Animated Button/C# and VB/Animated Button VB/Animated Button.vb
Imports System.Drawing
Imports System.Windows.Forms

Public Class AnimatedButton
    Inherits Button

    Private Class Frame
        Public Sub New(ByVal FrameCount As Integer, ByVal CurrentFrame As Integer, ByVal FrameSize As Size, ByVal ImageStrip As Image)

            Me.FrameCount = FrameCount
            Me.CurrentFrame = CurrentFrame
            Me.FrameSize = FrameSize
            Me.ImageStrip = ImageStrip

        End Sub

        Public ImageStrip As Image
        Public FrameCount As Integer
        Public CurrentFrame As Integer
        Public FrameSize As Size

    End Class

    Private _IsMouseOver As Boolean = False

    Private MouseOverFrame As Frame

    Private Property IsMouseOver As Boolean

        Get

            Return _IsMouseOver

        End Get

        Set(value As Boolean)

            _IsMouseOver = value

            Refresh()

        End Set

    End Property

    Protected Overrides Sub OnMouseEnter(e As EventArgs)

        MyBase.OnMouseEnter(e)

        IsMouseOver = True

    End Sub

    Protected Overrides Sub OnMouseLeave(e As EventArgs)

        MyBase.OnMouseLeave(e)

        IsMouseOver = False

    End Sub

    Private _IsMouseDown As Boolean = False

    Private Property IsMouseDown As Boolean

        Get

            Return _IsMouseDown

        End Get

        Set(value As Boolean)

            _IsMouseDown = value

            Refresh()

        End Set

    End Property

    Private MouseClickFrame As Frame

    Protected Overrides Sub OnMouseDown(mevent As MouseEventArgs)

        MyBase.OnMouseDown(mevent)

        IsMouseDown = True

    End Sub
    
    Protected Overrides Sub OnMouseUp(mevent As MouseEventArgs)

        MyBase.OnMouseUp(mevent)

        IsMouseDown = False

    End Sub

    Public Sub New()

        ' This call is required by the designer.
        InitializeComponent()

        ' Add any initialization after the InitializeComponent() call.

        DoubleBuffered = True

        MouseOverFrame = New Frame(25, 0, New Size(100, 30), My.Resources.ButtonMouseOver)
        MouseClickFrame = New Frame(25, 0, New Size(100, 30), My.Resources.ButtonMouseDown)

        Me.Size = MouseOverFrame.FrameSize

    End Sub

    Protected Overrides Sub OnPaint(ByVal e As System.Windows.Forms.PaintEventArgs)
        MyBase.OnPaint(e)

        'Add your custom paint code here

        e.Graphics.Clear(BackColor)

        e.Graphics.InterpolationMode = System.Drawing.Drawing2D.InterpolationMode.HighQualityBicubic
        e.Graphics.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality
        e.Graphics.CompositingQuality = System.Drawing.Drawing2D.CompositingQuality.HighQuality

        If (Not IsMouseOver AndAlso (Not IsMouseDown AndAlso (MouseOverFrame.CurrentFrame = 0))) Then

            e.Graphics.DrawImage(MouseOverFrame.ImageStrip,
                                 New Rectangle(0, 0, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height),
                                 New Rectangle(0, 0, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height),
                                 GraphicsUnit.Pixel)

        ElseIf (IsMouseOver AndAlso (Not IsMouseDown AndAlso (MouseOverFrame.CurrentFrame <> MouseOverFrame.FrameCount))) Then

            e.Graphics.DrawImage(MouseOverFrame.ImageStrip,
                                 New Rectangle(0, 0, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height),
                                 New Rectangle(0, MouseOverFrame.CurrentFrame * MouseOverFrame.FrameSize.Height, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height),
                                 GraphicsUnit.Pixel)

            MouseOverFrame.CurrentFrame += 1

        ElseIf (IsMouseOver AndAlso (Not IsMouseDown AndAlso ((MouseOverFrame.CurrentFrame = MouseOverFrame.FrameCount) AndAlso (MouseClickFrame.CurrentFrame = 0)))) Then

            e.Graphics.DrawImage(MouseOverFrame.ImageStrip,
                                 New Rectangle(0, 0, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height),
                                 New Rectangle(0, ((MouseOverFrame.CurrentFrame - 1) * MouseOverFrame.FrameSize.Height), MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height),
                                 GraphicsUnit.Pixel)

        ElseIf (IsMouseOver AndAlso (IsMouseDown AndAlso (MouseClickFrame.CurrentFrame <> MouseClickFrame.FrameCount))) Then

            e.Graphics.DrawImage(MouseClickFrame.ImageStrip,
                                 New Rectangle(0, 0, MouseClickFrame.FrameSize.Width, MouseClickFrame.FrameSize.Height),
                                 New Rectangle(0, MouseClickFrame.CurrentFrame * MouseClickFrame.FrameSize.Height, MouseClickFrame.FrameSize.Width, MouseClickFrame.FrameSize.Height),
                                 GraphicsUnit.Pixel)

            MouseClickFrame.CurrentFrame += 1

        ElseIf (IsMouseOver AndAlso (IsMouseDown AndAlso (MouseClickFrame.CurrentFrame = MouseClickFrame.FrameCount))) Then

            e.Graphics.DrawImage(MouseClickFrame.ImageStrip,
                                 New Rectangle(0, 0, MouseClickFrame.FrameSize.Width, MouseClickFrame.FrameSize.Height),
                                 New Rectangle(0, ((MouseClickFrame.CurrentFrame - 1) * MouseClickFrame.FrameSize.Height), MouseClickFrame.FrameSize.Width, MouseClickFrame.FrameSize.Height),
                                 GraphicsUnit.Pixel)

        ElseIf (IsMouseOver AndAlso (Not IsMouseDown AndAlso (MouseClickFrame.CurrentFrame <> 0))) Then

            MouseClickFrame.CurrentFrame -= 1

            e.Graphics.DrawImage(MouseClickFrame.ImageStrip,
                                 New Rectangle(0, 0, MouseClickFrame.FrameSize.Width, MouseClickFrame.FrameSize.Height),
                                 New Rectangle(0, MouseClickFrame.CurrentFrame * MouseClickFrame.FrameSize.Height, MouseClickFrame.FrameSize.Width, MouseClickFrame.FrameSize.Height),
                                 GraphicsUnit.Pixel)



        ElseIf (Not IsMouseOver AndAlso (Not IsMouseDown AndAlso (MouseOverFrame.CurrentFrame <> 0))) Then

            MouseOverFrame.CurrentFrame -= 1

            e.Graphics.DrawImage(MouseOverFrame.ImageStrip,
                                 New Rectangle(0, 0, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height),
                                 New Rectangle(0, MouseOverFrame.CurrentFrame * MouseOverFrame.FrameSize.Height, MouseOverFrame.FrameSize.Width, MouseOverFrame.FrameSize.Height),
                                 GraphicsUnit.Pixel)

        End If

        Using sf As New StringFormat

            sf.LineAlignment = StringAlignment.Center
            sf.Alignment = StringAlignment.Center

            e.Graphics.DrawString(Text, Font, New SolidBrush(ForeColor), e.ClipRectangle, sf)

        End Using

    End Sub

    Private Sub _FPS_Tick(sender As Object, e As EventArgs) Handles _FPS.Tick

        Refresh()

    End Sub
End Class
